### PR TITLE
Rework of input completions

### PR DESCRIPTION
**Pull request by r-stein**
This PR reworks the input-completion functionality. The main purpose was to add flexibility and user definable completions. It has been shortly discussed in https://github.com/SublimeText/LaTeXTools/issues/679 and will also solve https://github.com/SublimeText/LaTeXTools/issues/665.

A potential settings entry could be:

``` js
"fillall_helper_entries": [
	{
		"regex": "(?:\\][^{}\\[\\]]*\\[)?gvsedulcni\\\\",
		"extensions": ["svg"],
		"strip_extensions": [".svg"]
	},
],
```

Changes:

- completions (trigger regex, file extensions) are now defined as json compatible dicts and are also definable in the user settings
- a context has been added, such that is it possible to add completions without updating the keymap for `{`
- input completions can now optionally be triggered via `,` (e.g. for `\addbibresource`). (the key for this is `"comma_separated": True`)

TODO:
- ~~README entry~~
- [x] entry and description in default settings
- ~~additional options, (e.g. change the search folder)~~

### CODE REVIEW

**Comment by JimAmuro** [comment]
I'm sorry but still no luck for `\includesvg` and `\lstinput`...
I changed all the files here. No error message, it just writes `{}` in the file after I press `{`.

**Comment by r-stein** [comment]
@JimAmuro you could try to download or clone (`git clone https://github.com/r-stein/LaTeXTools -b fillall_helper_setting_entries`) the whole folder

File: LaTeXTools.sublime-settings [hunk 1]
```diff
@@ -282,6 +282,40 @@
 		]
 	},
 
+
+// ------------------------------------------------------------------
+// Additional fill all helper settings (for experts)
+// ------------------------------------------------------------------
+	// entries for the fill all helper
```
Comment (ig0774) [inline LaTeXTools.sublime-settings]: It'd be good to have some documentation on the format of entries expected... Not everyone may be willing to read the Python code.

**Comment by JimAmuro** [comment]
@r-stein I tried clone your repo today. Still no luck for entries in "fillall_helper_entries"... Windows 10. ST3. Deleted my user settings for LaTeXTools.

**Comment by r-stein** [comment]
@JimAmuro thats weird, you could look at the console `` ctrl+` `` output, when saving `latex_input_completions.py` and triggering `\includesvg{`

**Comment by JimAmuro** [comment]
@r-stein I have tried that, no error information.

**Comment by ig0774** [comment]
@r-stein: It might be a good idea to segregate out the fill-all helpers into their own file, separate from the main settings. At a technical level, there's no real justification for it, I just worry that this is going to make the already-intimidating settings file even worse.

**Comment by r-stein** [comment]
@ig0774 I am also struggeling with the complexity of the settings file. My idea would be to divide it into an advanced/expert settings file and normal settings file. My suggestion would be to add an additional settings file (`LaTeXToolsAdvanced.sublime-settings`), which is also accessable via the menu. If someone wants to overwrite an advances setting, he just copies the entry to his normal settings and changes it.
Hence the `get_settings` function would load `view.settings > LaTeXTools.sublime-settings > LaTeXToolsAdvanced.sublime-settings`.
Using this, one could move the cache or `open_image_command` setting to the advanced settings.

**Comment by ig0774** [comment]
@r-stein I suppose your proposal makes the most sense in terms of the simplest migration pathway. I'm a little hesitant about making `get_settings()` any more complex, but I suppose as long as we have a way to determine what the relevant setting is (which is part of the intention of #656), everything is ok.

**Comment by ig0774** [comment]
Shouldn't we also modify FillAll to leverage the `on_query_context` so it will be able to support dynamic entries as well as the static once?

**Comment by r-stein** [comment]
The `on_query_context` event listener supports both static and dynamic entries.

**Comment by ig0774** [comment]
@r-stein: Is this ready-to-go?

**Comment by r-stein** [comment]
No, I still want to add some documentation. Features and addition arguments (e.g. changing the search folder) could be a future PR.

File: latextools_utils/is_tex_file.py [hunk 2]
```diff
@@ -12,7 +12,6 @@
     strbase = str
 
 def get_tex_extensions():
-    view = sublime.active_window().active_view()
```
Comment (r-stein) [inline latextools_utils/is_tex_file.py]: I removed this, because it raises in error when called inside `plugin_loaded` in ST2 and is not used anyway.

**Comment by ig0774** [comment]
Documentation-wise, this might do well as a [wiki page](https://github.com/SublimeText/LaTeXTools/wiki). I'm not sure anyone actually reads them, but at least we'd have something.

**Comment by r-stein** [comment]
Agree that would be better.

**Pull request merged**